src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -0,0 +1,303 @@
+{
+	HASHCTL info;
+	uint32 hashFlags = 0;
+	HTAB *distributedTransactionHash = NULL;
+	int edgeIndex = 0;
+	int edgeCount = waitGraph->edgeCount;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(DistributedTransactionId);
+	info.entrysize = sizeof(TransactionNode);
+	info.hash = DistributedTransactionIdHash;
+	info.match = DistributedTransactionIdCompare;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT | HASH_COMPARE);
+
+	distributedTransactionHash =
+		hash_create("distributed deadlock detection", 64, &info, hashFlags);
+
+	for (edgeIndex = 0; edgeIndex < edgeCount; edgeIndex++)
+	{
+		WaitEdge *edge = &waitGraph->edges[edgeIndex];
+		TransactionNode *waitingTransaction = NULL;
+		TransactionNode *blockingTransaction = NULL;
+
+		DistributedTransactionId waitingId = {
+			edge->waitingNodeId,
+			edge->waitingTransactionNum,
+			edge->waitingTransactionStamp
+		};
+
+		DistributedTransactionId blockingId = {
+			edge->blockingNodeId,
+			edge->blockingTransactionNum,
+			edge->blockingTransactionStamp
+		};
+
+		waitingTransaction = GetTransactionNode(distributedTransactionHash, &waitingId);
+		blockingTransaction = GetTransactionNode(distributedTransactionHash, &blockingId);
+
+		waitingTransaction->waitsFor = lappend(waitingTransaction->waitsFor,
+											   blockingTransaction);
+	}
+
+	return distributedTransactionHash;
+}
+
+
+/*
+ * GetTransactionNode searches distributedTransactionHash for the given
+ * given transactionId. If the transaction is not found, the corresponding
+ * transactionNode is initialized to be NULL.
+ */
+static TransactionNode *
+GetTransactionNode(HTAB *distributedTransactionHash,
+				   DistributedTransactionId *transactionId)
+{
+	TransactionNode *transactionNode = NULL;
+	bool found = false;
+
+	transactionNode = (TransactionNode *) hash_search(distributedTransactionHash,
+													  transactionId,
+													  HASH_ENTER, &found);
+	if (!found)
+	{
+		transactionNode->waitsFor = NIL;
+	}
+
+	return transactionNode;
+}
+
+
+/*
+ * DistributedTransactionIdHash returns hashed value for a given distributed
+ * transaction id.
+ */
+static uint32
+DistributedTransactionIdHash(const void *key, Size keysize)
+{
+	DistributedTransactionId *entry = (DistributedTransactionId *) key;
+	uint32 hash = 0;
+
+	hash = hash_uint32(entry->initiatorNodeIdentifier);
+	hash = hash_combine(hash, hash_any((unsigned char *) &entry->transactionNumber,
+									   sizeof(int64)));
+	hash = hash_combine(hash, hash_any((unsigned char *) &entry->timestamp,
+									   sizeof(TimestampTz)));
+
+	return hash;
+}
+
+
+/*
+ * DistributedTransactionIdCompare compares DistributedTransactionId's a and b
+ * and returns -1 if a < b, 1 if a > b, 0 if they are equal.
+ *
+ * DistributedTransactionId are first compared by their timestamp, then transaction
+ * number, then node identifier.
+ */
+static int
+DistributedTransactionIdCompare(const void *a, const void *b, Size keysize)
```

Comment:
These could maybe go into a utility file, since I imagine we'll want to compare distributed transaction IDs in other places.